design/reboot-interface.md

```diff
@@ -0,0 +1,193 @@
+
+To provide a declarative interface for performing a reboot, a new
+BareMetalHostRebootRequest CRD will be created. The spec of this custom
+resource will contain only a reference to a BareMetalHost. The status will
+contain timestamps for when the Host was powered off and on again. Once the
+power off time is present in the status, the host has been stopped. Each
+RebootRequest will be owned by the Host it refers to, so that if the Host is
+deleted all associated RebootRequest CRs will be cleaned up.
+
+The actual power management will be performed by the Host controller. This is
+necessary to avoid race conditions by ensuring that the ``Online`` flag and any
+reboot requests are managed in the same place. The Host controller
+will watch RebootRequest objects and trigger a reconciliation of the linked
+Host when one appears or changes.
+
+For simplicity, we will try to avoid creating a separate controller for the
+RebootRequest at all, but instead perform all operations on it in the course of
+reconciling the BareMetalHost object that it references. However, it may yet
+prove necessary to also create a trivial controller for the purposes of basic
+housekeeping (e.g. setting the owner of the resource).
+
+The creation timestamp of the request resource will be interpreted as the time
+of the reboot request. Requests that predate the last booted time will be
+immediately marked as complete without taking any further action. The Host
+controller will timestamp all pending reboot requests as having completed the
+power off stage whenever any of the following occur:
+
+* The Host has been deprovisioned (in this case the ``lastPoweredOn`` field
+  will be empty)
+* The ``lastPoweredOn`` time is after the time of the reboot request
+* The ``poweredOn`` status is ``false``
+
+Whenever there are pending reboot requests, the Host controller will respond by
+powering off the Host (if it is not already) and timestamping all pending
+requests as having completed the power off stage. Power will then be restored
+only if/when the ``Online`` spec of the Host is true. Once Hosts are powered up
+again, any satisfied reboot requests will be timestamped as having completed
+the power on stage, with this timestamp matching the ``lastPoweredOn`` time of
+the Host.
+
+## Drawbacks
+
+This approach entails a whole new custom resource definition to maintain, and
+the instances of this resource must be managed by the Host controller, which
+adds a degree of complexity.
+
+If large numbers of requests are created but never deleted, searching for
+pending requests may start to consume significant resources.
+
+## Future Enhancements
+
+### Labelling
+
+Adding labels to partially completed and completed requests would allow us to
+exclude them when querying for pending requests. That would eliminate the
+performance concerns associated with allowing old requests to accumulate. This
+would be best performed by a RebootRequest-specific controller.
```

Comment:
Why would this be best?

I like targeted controllers, but there's also a real benefit to having just one controller that "owns" a resource. In this case, I'd default to letting the BMH controller own the RebootRequest resource as a sort of child, which includes any writes that reflect its status, unless there's a clear reason to have a second controller.

Reply:
It seemed like it would be easier to have an independent thing that just checks the status and adds the label if it's missing and the request is complete. Then the Host controller doesn't have to worry about the label, or about something else removing the label.
I guess the Host controller is going to get notified anyway though, so maybe it doesn't actually add any value.

Reply:
We can code whatever reconciliation we want for the reboot request and put that logic in either controller. I don't see an advantage to putting the label part in a dedicated controller.

Reply:
The only thing I can think of is if something adds the label to request that is not actually complete - a dedicated reboot request controller would simply remove the label again on reconcile, but the Host controller would ignore all resources with the label (it doesn't have to, but that is the whole point of having a label).
We're getting into extreme edge case territory with borderline malicious actors there though.

Reply:
Ah, I see, good point. We can keep that in mind if this became a problem we want to address. But I think we'll be better off by (eventually) implementing some kind of cleanup for old reboots and not having the operator use labels to identify which reboot resources have a particular state.